public/js/src/module/connection.js

```diff
@@ -44,26 +89,40 @@ function uploadRecord( record ) {
         return Promise.reject( e );
     }
 
-    batches.forEach( batch => {
-        batch.instanceId = record.instanceId;
-        batch.deprecatedId = record.deprecatedId;
-    } );
+    /** @type { Promise<UploadBatchResult[]> } */
+    let resultsPromise = Promise.resolve( [] );
+
+    /** @type { UploadBatchResult } */
+    let result;
 
     // Perform batch uploads sequentially for to avoid issues when connections are very poor and
     // a serious issue with ODK Aggregate (https://github.com/kobotoolbox/enketo-express/issues/400)
-    return batches.reduce( ( prevPromise, batch ) => prevPromise.then( () => _uploadBatch( batch ) ), Promise.resolve() )
+    return batches.reduce( ( prevPromise, batch ) => {
+        return prevPromise.then( results => {
+            return _uploadBatch( batch ).then( result => {
+                results.push( result );
+
+                return results;
+            } );
+        } );
+    }, resultsPromise )
         .then( results => {
             console.log( 'results of all batches submitted', results );
 
-            return results[ 0 ];
-        } );
+            result = results[ 0 ];
+
+            if ( options.isLastSaved ) {
+                return formCache.setLastSavedRecord( record.enketoId, record );
+            }
+        } )
+        .then( () => result );
```

Comment:
While this is clearly adding last-saved support, I also discovered adding initial test coverage that the original return Promise was always resolving `undefined` because these results were not actually being captured. That said it’s not clear why we would capture them all if only the first is relevant. Mostly I tried to stick close to the apparent intent.

Reply:
I'm not sure what you mean. But to clarify:

1. The original behavior in the `reduce` callback didn't accumulate results, it just stepped through each batch and returned the final `Promise<UploadBatchResult>`.
2. The next `then` refers to `results` and `results[ 0 ]` so it appeared the intent/expectation was that the `reduce` _was_ accumulating results, so that the _first_ batch's `UploadBatchResult` could be returned. Instead, `results[ 0 ]` was being coerced to `results[ '0' ]`, where the `UploadBatchResult` property did not have a `'0'` property, so it would always resolve to `undefined`.
3. While I was making the change here to support last-saved, I also addressed 1 & 2 by accumulating results as `Promise<Array<UploadBatchResult>>`, then resolving the first one, as that did appear to be the original intent.

Reply:
Thanks. That seems great.